enhancements/insights/insights-operator-runtime.md

```diff
@@ -0,0 +1,532 @@
+	// Hash of the version identifier of the Operating System (based on /etc/os-release VERSION_ID)
+	OsVersion string `json:"osVersion,omitempty"`
+	// Identifier of the kind of runtime
+	Kind string `json:"kind,omitempty"`
+	// Version of the kind of runtime
+	KindVersion string `json:"kindVersion,omitempty"`
+	// Entity that provides the runtime-kind implementation
+	KindImplementer string `json:"kindImplementer,omitempty"`
+	// Runtimes components
+	Runtimes []RuntimeComponent `json:"runtimes,omitempty"`
+}
+
+type RuntimeComponent struct {
+	// Name of a runtime used to run the application in the container
+	Name string `json:"name,omitempty"`
+	// The version of this runtime
+	Version string `json:"version,omitempty"`
+}
+```
+
+These new data elements are all optional (as they may not be relevant depending on the type of workloads) and are hashed (and truncated) to preserve anonymity (the hashing algorithm is [the same one that is currently used by the gatherer](https://github.com/openshift/insights-operator/blob/30b161f98eb185bb357ac84f8aca58e369d34bcd/pkg/gatherers/workloads/workloads_info.go#L336)).
+
+### Runtime Info Description
+
+With the ability to inspect running containers in OpenShift clusters, Red Hat can collect additional data about the software running on the cluster to improve its products.
+
+The additional data are:
+
+* Operating System (OS) information - Based on standard linux file ([/etc/os-release](https://www.freedesktop.org/software/systemd/man/latest/os-release.html)) present in most Linux distributions, collect the ID and Version ID of the operating system of the container.
+  * __`Os`__ - Identifier of the Operating System
+    * Raw value based on /etc/os-release `ID` field if present 
+    * Optional
+    * Value is a hash of the corresponding `ID` field: `debian, rhel, fedora, ol,...` (list not exhaustive)
+  * __`OsVersion`__ - Version identifier of the Operating system
+    * Raw value based on /etc/os-release `VERSION_ID` field if present 
+    * Optional
+    * Value is a hash of the corresponding `VERSION_ID` field: `11.0, 7.9, 17,..` (list not exhaustive)
+* Kind of runtime information - Based on the process name or ELF headers, determine the kind of runtime that is running:
+  * __`Kind`__ - Identifier of the kind of runtime
+    * Derived value (set by the insights-operator-runtime component)
+    * Optional (if detected by the insights-operator-runtime component)
+    * Value is a hash of the detected runtime kind: `Java, Golang, Node.js, GraalVM` (exhaustive list based on the capabilities of the container scanner)
+  * __`KindVersion`__ - Version of the kind of runtime.
+    * Raw value (depending on the value of runtime-kind, the version is read from different files in the container)
+    * Optional - based on the existence of various files in the containers (`$JAVA_HOME/release` for Java, the ELF header of the executable for Golang, the output of `node —version` for Node.js)
```

Comment:
nit: `—version` -> `--version`.

And is `node --version` run in the workload namespace?  I see some discussion below about fingerprint executables, but just because a command named `node` exists, doesn't mean it's necessarily the Node.js executable you think it might be, and there's an outside chance that it's a script where the `--version` option means "run `rm -rf /`" or something else that the workload admin might not appreciate.

On the other hand, trying to recreate a similar container in your own scratch space seems heavy, and like something that might create Event noise in the workload namespace (e.g. trying to launch your own Pod in their namespace to run the fingerprint command outside of their Pod) or fail to work in an Insights namespace (e.g. trying to pull an image that needs a namespace-specific pull secret).

Reply:
That's a risk and limitation of executing `cmd --version` in the workload namespace if the. `--version` has any undesirable side effect.
As we target some specific executable (`node`, `python`), we may instead look at a specific string in the executable to infer their version (for Node.js, something like `strings node | grep "node.js/v"`) but this is also brittle as we are looking at internals of the executable that are subject to change without notice.

Reply:
probably worth calling these out as in the risks and/or alternatives sections?

Reply:
I amended the RFE to call out that risk:

> ###### Risk of misbehaving code execution
> A few fingerprints relies on executing process in the user namespace (for example, executing `node --version` to determine the version of Node.js).
> There is a risk of side-effects if this execution does not match the expectation (if the `node` process is not the Node.js executable and its `--version` parameter triggers a `rm -rf /`).
Executing such code should be done conservatively on well-identified workloads (`python`, `node`, etc.) if they provide a way to get their versions without side-effect.